stan/math/rev/core/operator_addition.hpp

```diff
@@ -124,11 +124,8 @@ inline auto add(const VarMat1& a, const VarMat2& b) {
   arena_t<VarMat2> arena_b(b);
   arena_t<ret_type> ret(arena_a.val() + arena_b.val());
   reverse_pass_callback([ret, arena_a, arena_b]() mutable {
-    for (Eigen::Index i = 0; i < ret.size(); ++i) {
-      const auto ref_adj = ret.adj().coeffRef(i);
-      arena_a.adj().coeffRef(i) += ref_adj;
-      arena_b.adj().coeffRef(i) += ref_adj;
-    }
+    arena_a.adj() += ret.adj();
+    arena_b.adj() += ret.adj();
```

Comment:
Without this change, the adjoints that come back through:

```
auto x_sym = stan::math::eval(stan::math::multiply(x + x.transpose(), 0.5));
```

look funny. Like instead of the adjoints of x being:

```
1 2
3 4
```

you get:

```
1 5
0 4
```

Are you okay with this change? Or do you want do add another `operator+` specialization? Or is this a problem with transpose using views and we should change how transpose works?

Reply:
I thought about this some more and I'm wondering if the `reverse_pass_callbacks` that do linear access in memory might not work quite right if the inputs are view types (so like transposes, or blocks, or something strided)

Reply:
It might be worth looking at adopting a similar approach to the [read_var](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/read_var.hpp) and [read_fvar](https://github.com/stan-dev/math/blob/develop/stan/math/fwd/fun/read_fvar.hpp) approaches to copy values to two matrices in a single pass without indexing (using nullary functors)

Reply:
Oh that's really weird, @t4c1 shouldn't `arena_t<>` be evaluating any non-plain type matrices before putting them in the arena?

Reply:
@andrjohns that would look nice though I think the issue here is that an expression is not being evaluated and that's causing the adjoint values to goof

Reply:
Ah true true

Reply:
Let's not change this until we figure out what is happening

Reply:
Okay looking at multiply I think that's the actual problem. I'll come up with an example Friday :turkey:

Reply:
@bbbales2 can you give me a self-contained example I can go over with debugger?

Reply:
@t4c1 @SteveBronder here is a unit test:

```cpp
TEST(MathMixMatFun, testSumTranspose) {
  auto f = [](const auto& x) {
    return stan::math::eval(x + x.transpose());
  };

  Eigen::MatrixXd m(2, 2);
  m << 1, 2, 3, 4;

  stan::math::var_value<Eigen::MatrixXd> vm = m;
  Eigen::Matrix<stan::math::var, -1, -1> mv = m;

  auto out_vm = f(vm);
  auto out_mv = f(mv);

  out_vm.adj() << 0, 0, 1, 0;
  out_mv.adj() << 0, 0, 1, 0;
  stan::math::grad();

  std::cout << "Varmat:" << std::endl;
  std::cout << vm.adj() << std::endl;
  std::cout << "Matvar:" << std::endl;
  std::cout << mv.adj() << std::endl;
}
```

The output I get (if I revert this change to `operator+` is):
```
Varmat:
0 0
2 0
Matvar:
0 1
1 0
```

Reply:
I suspect this might also break with a block operation that takes a subset of a matrix.

If we want to do linear access for the autodiff we might have to add a bunch of stuff like over here: https://github.com/stan-dev/math/pull/2205/files#diff-9cdeb2ad158d329d95f157fedd9a23f8adf4b6b236e56d0c0c3463095f5637d8

Maybe a `require_eigen_linear_access_t` or something.

Reply:
You are right - the issue is in linear indexing. The code with loop and linear indexing assumes all matrices have the same memory layout. That is not true in this example - `.transpose()` return type is row-mayor, so its adjoint is not correct.

In this case we also can not use nullary functor - that also assumes same memory layout.

We need to either use separate expression evaluations (as Ben fixed this), two nested loops indexing with two indices, or special cases depending on whether `a`, `b` and `ret` all have the same memory layout.

Reply:
@SteveBronder @t4c1 So what is the next step?

Should I replace the linear index version here with matrix operations. Or add template logic so we can have a linear version and a separate matrix/array version? Or let me know if there's another option

Reply:
I think among the simple fixes available, you picked the best one. Two loops would likely be faster in many cases, but they would be very slow compared to this if both operands are row-major. If we want best possible performance for all cases there is a lot of separate cases to handle. I can try to make a general solution, but will take some time.

Reply:
Cool. Looks like @SteveBronder is fixing `operator` over here: https://github.com/stan-dev/math/pull/2223 so I'll just wait on that pull to finish this.

Reply:
#2223 got fixed